mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresMailboxMessageDAO.java

```diff
@@ -273,31 +270,55 @@ public Mono<Pair<Integer, Integer>> countTotalAndUnseenMessagesByMailboxId(Postg
     }
 
     public Flux<Pair<SimpleMailboxMessage.Builder, Record>> findMessagesByMailboxId(PostgresMailboxId mailboxId, Limit limit, MessageMapper.FetchType fetchType) {
-        PostgresMailboxMessageFetchStrategy fetchStrategy = FETCH_TYPE_TO_FETCH_STRATEGY.apply(fetchType);
-        Function<DSLContext, SelectSeekStep1<Record, Long>> queryWithoutLimit = dslContext -> dslContext.select(fetchStrategy.fetchFields())
-            .from(MESSAGES_JOIN_MAILBOX_MESSAGES_CONDITION_STEP)
-            .where(MAILBOX_ID.eq(mailboxId.asUuid()))
-            .orderBy(DEFAULT_SORT_ORDER_BY);
+        if (limit.isUnlimited()) {
+            return Flux.range(0, OFFSET_UNLIMITED)
+                .concatMap(offsetIndex -> findMessagesByMailboxIdBatch(mailboxId, fetchType, offsetIndex * queryBatchSize, queryBatchSize))
+                .takeUntil(List::isEmpty)
+                .flatMapIterable(Function.identity());
+        } else {
+            return findMessagesByMailboxIdBatch(mailboxId, fetchType, 0, limit.getLimit().get())
+                .flatMapIterable(Function.identity());
+        }
+    }
 
-        return postgresExecutor.executeRows(dslContext -> limit.getLimit()
-                .map(limitValue -> Flux.from(queryWithoutLimit.andThen(step -> step.limit(limitValue)).apply(dslContext)))
-                .orElse(Flux.from(queryWithoutLimit.apply(dslContext))))
-            .map(record -> Pair.of(fetchStrategy.toMessageBuilder().apply(record), record));
+    private Mono<List<Pair<SimpleMailboxMessage.Builder, Record>>> findMessagesByMailboxIdBatch(PostgresMailboxId mailboxId, MessageMapper.FetchType fetchType, int offset, int batchSize) {
+        PostgresMailboxMessageFetchStrategy fetchStrategy = FETCH_TYPE_TO_FETCH_STRATEGY.apply(fetchType);
+        return postgresExecutor.executeRows(dslContext -> Flux.from(dslContext.select(fetchStrategy.fetchFields())
+                .from(MESSAGES_JOIN_MAILBOX_MESSAGES_CONDITION_STEP)
+                .where(MAILBOX_ID.eq(mailboxId.asUuid()))
+                .orderBy(DEFAULT_SORT_ORDER_BY)
+                .limit(batchSize)
+                .offset(offset)))
```

Comment:
Actually we would be WAY more resilient not using an offset. We SHALL reuse the highest UID from the previous batch and resume from it. This would make us reslient to deletion in already paged items.

Reply:
Good idea
Should we also implement this for the Postgres BlobStore DAO?
We can add an additional column, such as a sequence or counter column?

Reply:
> Should we also implement this for the Postgres BlobStore DAO?

It would rely on alphabetic ordering, it is likely less important.